Writing-a-Language-Service-Plugin.md

```diff
@@ -50,13 +50,7 @@ function init(modules: {typescript: typeof ts_module}) {
 
     function create(info: ts.server.PluginCreateInfo) {
         // Set up decorator
-        const proxy = Object.create(null) as ts.LanguageService;
-        const oldLS = info.languageService;
-        for (const k in oldLS) {
-            (<any>proxy)[k] = function () {
-                return oldLS[k].apply(oldLS, arguments);
-            }
-        }
+        const proxy = {...info.languageService};
```

Comment:
this messes up the `this` reference.

Reply:
I thought it would, too, but it doesn't seem to. I tested before sending this PR, and also you can see for yourself in a browser console.

```js
const x = {f() { console.log('hi'); }, g() { this.f(); }};
x.g(); // prints 'hi'
const y = {...x};
y.g(); // also prints 'hi'
```

Regular classes don't work because of methods being on the prototype rather than the object, but this applies to the old technique as well, so it must not be how the language service is defined (I can't find where that is with a cursory search) and has nothing to do with `this`.

Reply:
well, this will work since `y` has a copy of all members of `x`, and `y` will be used as the `this` reference`. 
your code is a copy operation on the LS, where the original was just a proxy. this allows multiple proxies to have their own objects/own state and not step on each other's data.

Also this pattern allows you to capture the `oldLS`, say for instance you want to "augment" the completion list, so you first call `oldLS.getCompletionsAtPosition`, then add a few items to the result and send that back in your own `proxy.getCompletionsAtPosition` without having to worry about which object state lives on.

Reply:
When you say "own objects/own state", what do you mean? Augmenting `this` with new data? That seems like a poor practice to suggest, especially given that TS doesn't allow it by default (you could if you reopened the type, I guess). What state is there on the language service that shouldn't be considered common?

I'm also not sure why this approach differs in terms of capturing the old language service. This very guide does that, but doesn't actually use the oldLS variable to do so, so I didn't have to change anything.

I guess I would find it surprising if I had some method that I replaced, like `foo`, but whether my `foo` or the old `foo` was called depended on whether the editor was calling the method or whether some other method on the language service was.

Reply:
Changed it back to the old proxy, but without using 'any'. Let me know if you want me to revert it completely.

Looking at the old code, though, I'm even more convinced; there's no "state" stored on the language service anyway -- all of the properties are methods, no data. Otherwise the code wouldn't work, because it would call `.apply` on something other than a method.